### PR TITLE
Detect NIP-05 mentions

### PR DESCRIPTION
**Pull request by martindsq**
#923 

This PR basically does three things:

- Prevents the url parsing code from detecting links in emails, like server.com in martin@server.com
- Parses nip05 identifiers in notes, like linda@nos.social, and substituting that with a link to njump
- Adds the possibility of parsing bare npub identifiers with a @ prepended, like @npub1...

I added a concise way of testing this at the end of #923.

Another way of testing is composing a note with `Hello martin@nos.social` as the content and then see how Nos displays that

### CODE REVIEW

File: Nos/Extensions/String+Markdown.swift
Comment (mplorentz) [inline Nos/Extensions/String+Markdown.swift:21]: Can you add a comment here? Especially I am wondering after reading the function signature: what is it replacing the domains with?

File: Nos/Extensions/String+Markdown.swift
Comment (mplorentz) [inline Nos/Extensions/String+Markdown.swift:56]: can you add a comment here too?

**Review by mplorentz — DISMISSED** [review]
This looks good! It works perfectly and is better than what we had before. I really wish that tapping the NIP-05 opened the user's profile in our app instead of on njump. Especially since there is no easy way to get from njump back to our app yet. But I realize that would be more work. We'd probably need to show some kind of loading indicator after tapping the link while we check to see if it's a valid NIP-05. Maybe we could make a ticket for this as future work, or maybe it's not worth that. Maybe @setch-l can let us know.

File: Nos/Extensions/String+Markdown.swift
Comment (joshuatbrown) [inline Nos/Extensions/String+Markdown.swift:56]: Same as above -- this doesn't need to be in a String extension since it's not using `self`.

**Review by joshuatbrown — CHANGES_REQUESTED** [review]
Looks good overall; some suggestions and a question.

File: NosTests/NoteParserTests.swift
Comment (martindsq) [reply NosTests/NoteParserTests.swift:59]: We need to define how we want to achieve that, if we want to.

I think we won't want to be doing requests when parsing notes, which is what we need to do in order to know which is the npub. That leave us with two alternatives:

1. Modify ProfileView so that it can be shown with just a NIP-05 (that will first request the npub and then look if we have the profile locally)
2. Don't do anything, just bare NIP-05 identifiers are not good ways of referencing to people in nostr, for example, because it doesn't notify those users you are mentioning them. We should be encouraging users to mention them accordingly in the editor instead of just writing martin@nos.social.

So, as I think any of these alternatives is outside the scope of this ticket, which was just fixing the display issues, I decided to just link to jump as a middle ground alternative that didn't cost me much effort.

**Comment by martindsq** [comment]
> This looks good! It works perfectly and is better than what we had before. I really wish that tapping the NIP-05 opened the user's profile in our app instead of on njump. Especially since there is no easy way to get from njump back to our app yet. But I realize that would be more work. We'd probably need to show some kind of loading indicator after tapping the link while we check to see if it's a valid NIP-05. Maybe we could make a ticket for this as future work, or maybe it's not worth that. Maybe @setch-l can let us know.

@mplorentz please, check my answer to one of Josh' comments regarding this.

File: NosTests/NoteParserTests.swift
Comment (mplorentz) [reply NosTests/NoteParserTests.swift:59]: I see a third way:
1. make all NIP-05s links to a custom URL scheme (nos://nip05/martin@nos.social)
2. Handle this URL scheme in `Router.open(url: URL, with context: NSManagedObjectContext)`
3. Create an observable property like `isLoading` on the `Router`.
4. Use it to show a loading indicator on `AppView` while we look up the NIP-05 (disable UI behind it)
5. If it's a valid NIP-05 and we get an npub, open the profile for that npub. If not, this might be an email address, open the mail app.

This avoids us trying to make web requests when formatting the note contents, at the cost of a little loading time when the user taps. It also solves the problem of email addresses being mistaken for NIP-05s.

**Comment by setch-l** [comment]
@mplorentz @martindsq - I put something in the next sprint to work out user journeys for clicking on all the n things. Sebastian and I will be in touch.

File: NosTests/NoteParserTests.swift
Comment (martindsq) [reply NosTests/NoteParserTests.swift:59]: @mplorentz that would be option 1 in my previous message, we don't need to use the url scheme I we don't want it, we can just open the ProfileView with the same mechanism me open it by following links to @npub1 or MessageView by following links to %note1. I will create a different ticket for this, so we can get these bugfixes I'm adding here in Production until we work on following the NIP-05 identifiers the upcoming sprint.

**Comment by martindsq** [comment]
@joshuatbrown  @mplorentz I did some refactoring to address some comments Josh made (and I completely agree with).

I switched `NoteParser` from `enum` to `struct`, replaced all static functions with just functions and updated the tests accordingly (with a sut object).

I refactored extractURLs to avoid referencing `self`, and then moved it, and its three helper functions, to a new URLParser object with just extractURL as the public function so NoteParser can use it (it was the only client of those functions) and we can test it. I created a URLParserTests file and moved the unit tests there.

I also did the same with the remaining function in that extension to URLParser.

And finally I renamed `extractURLs` to `replaceUnformattedURLs` and the remaining function I just mentioned to `findUnformattedURLs` as I think both together make more sense with those names.

File: NosTests/NoteParserTests.swift
Comment (martindsq) [reply NosTests/NoteParserTests.swift:59]: I created an issue here https://github.com/planetary-social/nos/issues/1087 and moved it to this sprint so I work on it in a different branch. I think the njump links, while a bit weird, are functional and we can have this on production until we finish that issue.

Comment (joshuatbrown) [reply NosTests/NoteParserTests.swift:59]: This makes sense to me and I'm good with it.

File: Nos/Views/NoteOptionsButton.swift
Comment (joshuatbrown) [reply Nos/Views/NoteOptionsButton.swift:56]: @martindsq still hoping we can make this change; maybe you just haven't gotten to it yet.

**Review by joshuatbrown — CHANGES_REQUESTED** [review]
I've tested this and everything works as expected. I'm still hoping for a few small changes as suggested previously, but this is great!

File: Nos/Views/NoteOptionsButton.swift
Comment (martindsq) [reply Nos/Views/NoteOptionsButton.swift:56]: I did change it. Let me check why it is not here yet.

**Comment by martindsq** [comment]
@joshuatbrown I hadn't staged the changes :( Sorry about that. Please, check this again.

File: Nos/Models/CoreData/Event+CoreDataClass.swift
Comment (joshuatbrown) [inline Nos/Models/CoreData/Event+CoreDataClass.swift:918]: Not 100% sure how practical it is to test this, but... I'd probably make a `NoteParser` parameter for this func, even if it just defaults to `NoteParser()`. That would make it testable.

File: Nos/Models/CoreData/Event+CoreDataClass.swift
Comment (joshuatbrown) [inline Nos/Models/CoreData/Event+CoreDataClass.swift:969]: Same as above -- a `NoteParser` parameter would be slightly nicer.

File: Nos/Views/GoldenPostView.swift
Comment (joshuatbrown) [inline Nos/Views/GoldenPostView.swift:80]: Maybe this URLParser could be a property on GoldenPostView -- maybe even a `@Dependency`? (I'm still not sure how to make it available as an `@Dependency` but seems like it should be possible.)

File: Nos/Views/GoldenPostView.swift
Comment (joshuatbrown) [inline Nos/Views/GoldenPostView.swift:89]: Same as above -- maybe a dependency property on GoldenPostView.

File: NosTests/NoteParserTests.swift
Comment (joshuatbrown) [inline NosTests/NoteParserTests.swift:5]: One thing we might do to try to keep this file length below our threshold is to move the `AttributedString` extension to another file. Not sure if we have anything else like that we could move out, but I'd love it if we could keep files smallish.

Maybe even a new file that contains an `extension NoteParserTests` would help, and it could contain a set of tests that are all related to each other somehow (events, like `testContentWithUntaggedEvent` and similar?).

File: NosTests/NoteParserTests.swift
Comment (joshuatbrown) [inline NosTests/NoteParserTests.swift:7]:
```suggestion
final class NoteParserTests: CoreDataTestCase {
```

**Review by joshuatbrown — CHANGES_REQUESTED** [review]
Looking really good! I have a few more thoughts on the implementation but this is really close and almost ready to merge from my perspective.

**Comment by martindsq** [comment]
@joshuatbrown I created two extensions to NoteParserTests, one for those that test NIP-08 mentions (I mean, correct mentions), and another for NIP-27 mentions. All the other tests, test mostly stuff like parsing unconventional mentions, like just pasting an npub, that is, they do not follow protocol conventions.

File: Nos/Views/GoldenPostView.swift
Comment (martindsq) [reply Nos/Views/GoldenPostView.swift]: Nice, I added it as a dependency.

File: NosTests/NoteParserTests.swift
Comment (martindsq) [reply NosTests/NoteParserTests.swift]: See my comment below.

File: NosTests/NoteParserTests.swift
Comment (mplorentz) [reply NosTests/NoteParserTests.swift:66]: Please check with Linda on the priority of #1087 if you haven't already.

File: Nos/Models/NoteParser.swift
Comment (mplorentz) [inline Nos/Models/NoteParser.swift:7]: @joshuatbrown @martindsq For some reason I can't find the comment about this change. Can you link it to me or explain why it was made? Each of the functions below are pure, and I find it a best practice to make pure functions `static` because it is a clear signal to the developer: "this function doesn't use self" If you want to collect a bunch of related pure functions together it makes sense to do so in a stateless `enum`, because you can't instantiate a stateless `enum`. This is a pattern we use across the codebase. The `struct` here allows us to instantiate a `NoteParser`, but why? NoteParser doesn't need to store anything in memory.

Comment (martindsq) [reply Nos/Models/NoteParser.swift:7]: The discussion that sparked this change is this one https://github.com/planetary-social/nos/pull/1043#discussion_r1569388861. At the time of adding NoteParser, changing it ti enum was also one of your suggestions https://github.com/planetary-social/nos/pull/281#discussion_r1181859468.

1. `enum` are for enumerations and that's it, if we use it in a different and unique way, that will confuse the clients of `NoteParser` because the signature will be semantically incongruent. 

2. I think the clients of NoteParser should not care about if the functions inside use `self` or not. That's specific to the implementation. The best practice here, taking into account Design Patterns, is to encapsulate a complex algorithm in an object and hide the complexity in an interface. NoteParser doesn't need to store anything in memory, but it could later, it could start having parameters the user could tune in the UI, or we could be adding some private vars to make the code easier to read. And that kind of change in the implementation of NoteParser won't affect all clients of NoteParser if it was already an object, but it will if it was a bunch of static functions before.

Comment (mplorentz) [reply Nos/Models/NoteParser.swift:7]: Hm we are definitely in disagreement here. 

> enum are for enumerations and that's it

Using caseless enums is a fairly common pattern in Swift in my experience, in fact it's in the default swiftlint rules. If you create a struct without all static functions it will warn you to change it to a caseless enum. It's true, a caseless enum is not an enumeration which might be confusing the first time you see it, but you actually _can't misuse` it by instantiating one which is the point of its use.

Also I see from the linked comment that both you and @joshuatbrown prefer not to use static functions. Can you explain more why that is? Part of my development philosophy is to make types as strict as you can. For instance, I wouldn't want to pass a string to a function if you can instead use an enum. Similarly a static function is a stricter type of function, it is not attached to a class or value so (barring mutation of global variables or other no-nos) you know that it will clean up all its memory when the function ends. 

If I want to parse a note and I'm seeing in autocomplete that I can just call `NoteParser.parse(...)` then I know I can just autocomplete that and I'm good to go. If I see that I need to instantiate a NoteParser now I need to answer a bunch of further questions to use responsibly: who should own the NoteParser? How long does it need to stay allocated? What are the implications of calling `parse(...)` on its internal state? Of course these aren't that hard to answer by reading the code, but it's worse API design because I don't need to read the code in the first example.

I know that sometimes people want something they can instantiate so they can mock certain functions for a test. In that case I think it's best to use a `class` and mark all the static functions `class func`, but we aren't doing any mocking for NoteParser so I don't think that applies.

(Also feel free to merge this if Josh approves. I do want to finish this discussion but I don't want it to hold this feature from shipping)

Comment (martindsq) [reply Nos/Models/NoteParser.swift:7]: I agree that we should be merging this if @joshuatbrown given that we are discussing the style here and whatever we decide won't impact the functionality. Josh, mind reviewing again?

Regarding the discussion, it's hard to separate what is good and wrong from what is just a matter of style. I don't think `static` functions are stricter, they could be modifying `static` properties; for me, `static` functions are more commonly used for constructing objects or small functions meant to be shared, and not for coding behavior or algorithms, these should be encapsulated in objects, that's how I learned it by reading the Design Patterns book. 

I think the doubts you have when instantiating a NoteParser are a bit unfair. For example, when I use a DateFormatter instance I don't need to answer complex questions like: who should own the DateFormatter? How long does it need to stay allocated? or What are the implications of calling a function on its internal state?


Also, the way we had this before was misleading. We had different static functions, but in reality, most of them were used internally by the other (the main one) and were meant to be modified together because the implementation was very specific to the problem we wanted to solve. I couldn't just reuse a function to solve something different. I think we should hide all this in an object. I agree it could be a `class` though, but I don't think `static` functions give us an edge, and make future modifications harder to make.

Comment (joshuatbrown) [reply Nos/Models/NoteParser.swift:7]: I think I agree with a lot of what both of you are saying.

> I find it a best practice to make pure functions static because it is a clear signal to the developer: "this function doesn't use self"

💯 

> If you want to collect a bunch of related pure functions together it makes sense to do so in a stateless enum, because you can't instantiate a stateless enum.

💯 

> I think the clients of NoteParser should not care about if the functions inside use self or not.

💯 

> I know that sometimes people want something they can instantiate so they can mock certain functions for a test.

🖐️ That's me!

> In that case I think it's best to use a class and mark all the static functions class func, but we aren't doing any mocking for NoteParser so I don't think that applies.

🎯 This is the key problem for me. We aren't doing any mocking for NoteParser. In order to test anything that uses NoteParser, like Event+CoreDataClass, we'd be testing NoteParser along with it. I'd rather test Event+CoreDataClass in isolation; this makes it (a) easier write tests for it, since less code is under test, and (b) easier to identify bugs in it, since we can eliminated NoteParser as the potential source.

If we could find a way to both make our code as strict as possible _and_ make it easy to isolate our code for testing, I'd be all for that. If we can't, I always prioritize testability.

File: Nos/Views/GoldenPostView.swift
Comment (joshuatbrown) [inline Nos/Views/GoldenPostView.swift:88]: This is a minor style issue, but for future PRs I'd prefer this:

```suggestion
        guard let content = note.content else { return nil }
```

This would eliminate the need for the `if...else` below and I think it's a little easier to read.

File: Nos/Views/New Note/NewNoteView.swift
Comment (joshuatbrown) [inline Nos/Views/New Note/NewNoteView.swift:212]: Again, I'd prefer this dependency on NoteParser to be declared up top as a `@Dependency` property rather than instantiated here inside a func.

File: NosTests/NosUITests/NosUITests.swift
Comment (joshuatbrown) [inline NosTests/NosUITests/NosUITests.swift:1]: I don't think these should be moved into the NosTest folder. That one contains everything in the NosTests target; this and the next file are in the NosUITests target.

File: NosTests/NoteParserTests+NIP08.swift
Comment (joshuatbrown) [inline NosTests/NoteParserTests+NIP08.swift:1]: 👍

**Review by joshuatbrown — CHANGES_REQUESTED** [review]
My main concern here, I think, is moving the UI tests into the NosTests folder. The other suggestions are minor and wouldn't block this PR from my perspective.

File: NosTests/NosUITests/NosUITests.swift
Comment (martindsq) [reply NosTests/NosUITests/NosUITests.swift:1]: I removed those I deleted the target (we weren't using it)

File: Nos/Views/New Note/NewNoteView.swift
Comment (martindsq) [reply Nos/Views/New Note/NewNoteView.swift]: I added a dependency just in the same way we added for URLParser

File: Nos/Views/GoldenPostView.swift
Comment (martindsq) [reply Nos/Views/GoldenPostView.swift]: Nice catch. I also prefer that. I think it was a consequence of moving that code from the `body` without realizing I could rewrite it.

**Review by joshuatbrown — APPROVED** [review]
Excellent! Thanks for all your work on this!

**Pull request merged**
